packages/pkh-webauthn/src/utils.ts

```diff
@@ -0,0 +1,254 @@
+ * @param {Uint8Array|ArrayBuffer} attestationObject As given by credentials.create().response.attestationObject
+ */
+export function decodeAuthenticatorData (authData: Uint8Array) {
+  authData = assertU8(authData)
+  // https://w3c.github.io/webauthn/#sctn-authenticator-data
+  if (authData.length < 37) throw new Error('AuthenticatorDataTooShort')
+  let o = 0
+  const rpidHash = authData.slice(o, o += 32) // SHA-256 hash of rp.id
+
+  const flags = authData[o++]
+  // console.debug(`Flags: 0b` + flags.toString(2).padStart(8, '0'))
+  if (!(flags & (1 << 6))) throw new Error('AuthenticatorData has no Key')
+
+  const view = new DataView(authData.buffer)
+  const signCounter = view.getUint32(o); o += 4
+
+  // https://w3c.github.io/webauthn/#sctn-attested-credential-data
+  const aaguid = authData.slice(o, o += 16)
+  const clen = view.getUint16(o); o += 2
+  const credentialId = authData.slice(o, o += clen)
+
+  // https://datatracker.ietf.org/doc/html/rfc9052#section-7
+  // const publicKey = decode(authData.slice(o)) // cborg.decode fails; Refuses to decode COSE use of numerical keys
+  const cose = decodeCBORHack(authData.slice(o)) // Decode cbor manually
+
+  // Section 'COSE Key Type Parameters'
+  // https://www.iana.org/assignments/cose/cose.xhtml
+  if (cose[1] !== 2) throw new Error('Expected EC Coordinate pair')
+  if (cose[3] !== -7) throw new Error('Expected ES256 Algorithm')
+  const x = cose[-2]
+  const y = cose[-3]
+  if (!(x instanceof Uint8Array) || !(y instanceof Uint8Array)) throw new Error('Expected X and Y coordinate to be buffers')
+  const publicKey = ecPointCompress(x, y)
+  return {
+    rpidHash,
+    flags,
+    signCounter,
+    aaguid,
+    credentialId,
+    publicKey,
+    cose
+  }
+}
+
+/**
+ * Normalize authenticatorData across browsers/runtimes
+ * different runtimes implement different parts of spec.
+ */
+export function getAuthenticatorData (response: any) {
+  if (response.getAuthenticatorData === 'function') return response.getAuthenticatorData() // only on Chrome
+  if (response.authenticatorData) return response.authenticatorData // Sometimes not available on FF
+  if (response.attestationObject) { // Worst case scenario, decode attestationObject
+    const { authData } = decode(assertU8(response.attestationObject))
+    return assertU8(authData)
+  }
+  throw new Error('Failed to recover authenticator data from credential response') // Give up
+}
+
+/**
+ * Normalize ArrayBuffer|Uint8Array => Uint8Array or throw
+ */
+function assertU8 (o: Uint8Array | ArrayBuffer) : Uint8Array {
+  if (o instanceof ArrayBuffer) return new Uint8Array(o)
+  if (o instanceof Uint8Array) return o
+  throw new Error('Expected Uint8Array')
+}
+
+/**
+ * Tiny unsafe CBOR decoder that supports COSE_key numerical keys
+ * https://www.iana.org/assignments/cose/cose.xhtml
+ * Section 'COSE Key Type Parameters'
+ * TODO: check if iso-webauthn package handles this
+ */
+function decodeCBORHack (buf: Uint8Array) {
```

Comment:
I feel a bit bad about merging awesome-sauce decoder, it does not support full CBOR spec and will throw errors if used on anything other than COSE objects.. what should I do?

Reply:
On further inspection, i was wrong. package `cborg` does support negative integer keys when option `{ useMaps: true }` is passed, but in one of the testcases it lost the track of offsets and failed altogether to decode the COSE message.

https://github.com/rvagg/cborg/issues/92

There is the `iso-passkeys` library, but it uses a different eco of dependencies and attempts to support more algorithms than the `ES256 / -7`.